### PR TITLE
Rudimentary codelist support

### PR DESCRIPTION
**Pull request by evansd**
This adds basic support for querying using codes. We add individual classes for each coding system. This allows us to use the type system to enforce that the right coding system is used.

There's obviously lots more to be done here, but I'd appreciate feedback on the general approach before we go too much further with it.

### CODE REVIEW

File: databuilder/query_language.py [hunk 1]
```diff
@@ -54,6 +55,11 @@ def is_null(self):
         return _apply(qm.Function.IsNull, self)
 
     def is_in(self, other):
+        # The query model requires a Set type (and an immutable one at that) for
+        # containment queries, but that's a bit user-unfriendly so we accept
+        # lists/tuples etc here and convert them
+        if isinstance(other, Iterable) and not isinstance(other, (str, bytes)):
+            other = frozenset(other)
```
[the remark this answers is not in the document]
Comment (benbc) [reply databuilder/query_language.py:63]: I do think that loosening the types accepted is a good idea.

File: databuilder/query_engines/sqlite.py [hunk 2]
```diff
@@ -80,7 +82,15 @@ def get_sql(self, node):
 
     @get_sql.register(Value)
     def get_sql_value(self, node):
-        return sqlalchemy.literal(node.value)
+        return sqlalchemy.literal(self.convert_value(node.value))
+
+    def convert_value(self, value):
+        if isinstance(value, BaseCode):
+            return value.value
```
Comment (benbc) [inline databuilder/query_engines/sqlite.py:89]: Another alternative to mentioning `BaseCode` here would be to maintain a registry of mappings that was populated from outside.

Comment (benbc) [reply databuilder/query_engines/sqlite.py:89]: I don't feel strongly which of the three options you take, but we should keep an eye on this and consider refactoring later.

File: databuilder/query_language.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (evansd) [reply databuilder/query_language.py:61]: You're quite right. I'm not sure what I was thinking here. I have amended.

File: databuilder/query_engines/sqlite.py (hunk 2 above)
Comment (evansd) [reply databuilder/query_engines/sqlite.py:89]: OK, that sounds sensible. I just wanted to make sure there wasn't some Obviously Correct approach I was missing.

**Review by inglesp — APPROVED** [review]

**Pull request merged**